Docker Health Checks

luketainton opened the pull request:
Fix for #4954.

Added health checks to the Wekan and MongoDB containers. This brings the following benefits:
1. The containers can auto-restart on crash
2. The Wekan app container will not start until MongoDB is ready

**These changes work for me, but so did my last PR. Please double check that it works for you before merging!**

luketainton commented:
@mfilser Can you try this one please before merging?

mfilser commented:
@luketainton
It seems to work now in general, but starting the Docker container takes now longer, and stopping a lot longer. Maybe there are some settings to adjust?

luketainton commented:
It will take longer to start, because it's now waiting for MongoDB to be up and ready before it starts the application. 

As for it taking longer to shut down, I'll look into that tomorrow.

mfilser commented:
Is it important to wait for MongoDB to starr the app? I never had problems with starting the app and MongoDB parallel.

luketainton commented:
I don't know for sure, I just figured it was best practice. I could comment it out in the Compose file with a comment explaining what it does if that fits better?

mfilser commented:
Just thinking. What was your reason to add healthcheck to Wekan? Because of "best practise" ?

luketainton commented:
I added one to Wekan itself because every so often my own instance seems to throw an HTTP 500 or something and I have to restart it. I added one to MongoDB while I was at it.

mfilser commented:
I am running Wekan with Docker since many years without any app crashes.

xet7 commented:
@luketainton @mfilser 

1. What operating system version you are using to run Docker?
2. How much total and free RAM that WeKan server has? It server swapping to disk?
3. How many CPU cores that WeKan server has? At Linux/Mac, command `nproc`. At Windows, Task Manager CPU info.
4. How many users are using WeKan at the same time?

xet7 commented:
For example, at CentOS 7 there is this issue:
https://github.com/wekan/wekan/blob/master/docker-compose.yml#L130-L136

xet7 commented:
Docker does crash if there is not enough RAM.

xet7 commented:
@luketainton @mfilser

What Docker-like software version are you using? For example, from www.docker.com , podman, or something other.

xet7 commented:
@luketainton 

1. If you set `DEBUG=true` at  `docker-compose.yml` , do you get better logs when it crashes?
2. When WeKan crashes, is there some error message at `/var/log/syslog` ?
3. Is there errors at `docker logs wekan.wekan` and `docker logs wekan.mongodb` ?
4. Is there any errors at browser right click Inspect Console? For example, are there errors about websockets? Webserver config for websockets etc for  Caddy/Nginx/Apache etc is at right menu of https://github.com/wekan/wekan/wiki

mfilser commented:
> @luketainton @mfilser
> 
>     1. What operating system version you are using to run Docker?
> 
>     2. How much total and free RAM that WeKan server has? It server swapping to disk?
> 
>     3. How many CPU cores that WeKan server has? At Linux/Mac, command `nproc`. At Windows, Task Manager CPU info.
> 
>     4. How many users are using WeKan at the same time?

I am running my main instance on an "AMD EPYC 7702P 64-Core Processor", 4 cores, 8 GB Ram the whole Server + SSD drives.
The Server before had a "Intel Gold CPU" (don't know which model, 2 cores, 8 GB Ram the whole Server + SAS drives, and on this Server i also didn't have any problems.

My current Docker version: 24.0.2, docker-compose: 2.13.0
My OS: Debian 11
Current Users: About 5 at the same time.

luketainton commented:
@xet7 sorry for delay, here's my responses:

Q: What operating system version you are using to run Docker?
A: It's running on my Synology NAS.

Q: How much total and free RAM that WeKan server has? It server swapping to disk?
A: Whole NAS has 6GB RAM. I'm not restricting how much WeKan can access.

Q: How many CPU cores that WeKan server has? At Linux/Mac, command nproc. At Windows, Task Manager CPU info.
A: INTEL Celeron J4025, 2x cores.

Q: How many users are using WeKan at the same time?
A: One.

Q: What Docker-like software version are you using? For example, from [www.docker.com](http://www.docker.com/) , podman, or something other.
A: Docker.

Q: If you set DEBUG=true at docker-compose.yml , do you get better logs when it crashes?
A: Not tried, got logs from Docker container. Can try if needed.

Q: When WeKan crashes, is there some error message at /var/log/syslog ?
A: Not checked, got logs from Docker container. Can check if needed.

Q: Is there errors at docker logs wekan.wekan and docker logs wekan.mongodb ?
A: Docker logs for app container show this:
```
Exception while polling query {"collectionName":"__pre_attachments","selector":{},"options":{"transform":null,"projection":{"_id":1,"isFinished":1}}} MongoTopologyClosedError: Topology is closed
    at processWaitQueue (/build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/sdam/topology.js:550:46)
    at Topology.selectServer (/build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/sdam/topology.js:297:9)
    at Topology.<anonymous> (/build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/sdam/topology.js:43:94)
    at internal/util.js:341:30
    at new Promise (<anonymous>)
    at Topology.selectServerAsync (internal/util.js:340:12)
    at executeOperationAsync (/build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/operations/execute_operation.js:39:24)
    at /build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/operations/execute_operation.js:12:45
    at maybeCallback (/build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/utils.js:337:21)
    at executeOperation (/build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/operations/execute_operation.js:12:38)
    at FindCursor._initialize (/build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/cursor/find_cursor.js:54:50)
    at FindCursor.[kInit] (/build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/cursor/abstract_cursor.js:444:14)
    at next (/build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/cursor/abstract_cursor.js:512:22)
    at internal/util.js:341:30
    at new Promise (<anonymous>)
    at next (internal/util.js:340:12)
    at /build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/cursor/abstract_cursor.js:228:20
    at maybeCallback (/build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/utils.js:337:21)
    at FindCursor.next (/build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/cursor/abstract_cursor.js:224:42)
    at packages/mongo/mongo_driver.js:1118:22
    at new Promise (<anonymous>)
    at SynchronousCursor._rawNextObjectPromise (packages/mongo/mongo_driver.js:1117:12)
 => awaited here:
    at Function.Promise.await (/build/programs/server/npm/node_modules/meteor/promise/node_modules/meteor-promise/promise_server.js:56:12)
    at packages/mongo/mongo_driver.js:1134:14
    at /build/programs/server/npm/node_modules/meteor/promise/node_modules/meteor-promise/fiber_pool.js:43:40
 => awaited here:
    at Promise.await (/build/programs/server/npm/node_modules/meteor/promise/node_modules/meteor-promise/promise_server.js:60:12)
    at SynchronousCursor._nextObject (packages/mongo/mongo_driver.js:1183:38)
    at SynchronousCursor.forEach (packages/mongo/mongo_driver.js:1198:22)
    at SynchronousCursor.getRawObjects (packages/mongo/mongo_driver.js:1248:12)
    at PollingObserveDriver._pollMongo (packages/mongo/polling_observe_driver.js:156:44)
    at Object.task (packages/mongo/polling_observe_driver.js:93:12)
    at Meteor._SynchronousQueue.SQp._run (packages/meteor.js:1026:16)
    at packages/meteor.js:1003:12 {
  [Symbol(errorLabels)]: Set(0) {}
}
```
Restarting the app container (not MongoDB) seem to fix this.

Q: Is there any errors at browser right click Inspect Console? For example, are there errors about websockets? Webserver config for websockets etc for Caddy/Nginx/Apache etc is at right menu of https://github.com/wekan/wekan/wiki
A: No, just an HTTP 502 Bad Gateway error.

xet7 commented:
Sorry this slows down starting WeKan.
